les/lespay/client/valuetracker.go

```diff
@@ -0,0 +1,511 @@
+	vt.lock.Lock()
+	defer vt.lock.Unlock()
+
+	nv.lastReqCosts = reqCosts
+	nv.updateCosts(reqCosts, &vt.refBasket.reqValues, vt.refBasket.reqValueFactor(reqCosts))
+}
+
+// RtStats returns the global response time distribution statistics
+func (vt *ValueTracker) RtStats() ResponseTimeStats {
+	vt.lock.Lock()
+	defer vt.lock.Unlock()
+
+	vt.periodicUpdate()
+	return vt.rtStats
+}
+
+// periodicUpdate transfers individual node data to the global statistics, normalizes
+// the reference basket and updates request values. The global state is also saved to
+// the database with each update.
+func (vt *ValueTracker) periodicUpdate() {
+	now := vt.clock.Now()
+	vt.statsExpLock.Lock()
+	vt.statsExpFactor = utils.ExpFactor(vt.statsExpirer.LogOffset(now))
+	vt.statsExpLock.Unlock()
+
+	for _, nv := range vt.connected {
+		basket, rtStats := nv.transferStats(now, vt.transferRate)
+		vt.refBasket.add(basket)
+		vt.rtStats.AddStats(&rtStats)
+	}
+	vt.refBasket.normalize()
+	vt.refBasket.updateReqValues()
+	for _, nv := range vt.connected {
+		nv.updateCosts(nv.lastReqCosts, &vt.refBasket.reqValues, vt.refBasket.reqValueFactor(nv.lastReqCosts))
+	}
+	vt.saveToDb()
+}
+
+type ServedRequest struct {
+	ReqType, Amount uint32
+}
+
+// Served adds a served request to the node's statistics. An actual request may be composed
+// of one or more request types (service vector indices).
+func (vt *ValueTracker) Served(nv *NodeValueTracker, reqs []ServedRequest, respTime time.Duration) {
+	vt.statsExpLock.RLock()
+	expFactor := vt.statsExpFactor
+	vt.statsExpLock.RUnlock()
+
+	nv.lock.Lock()
+	defer nv.lock.Unlock()
+
+	var value float64
+	for _, r := range reqs {
+		nv.basket.add(r.ReqType, r.Amount, nv.reqCosts[r.ReqType]*uint64(r.Amount), expFactor)
+		value += (*nv.reqValues)[r.ReqType] * float64(r.Amount)
+	}
+	nv.rtStats.Add(respTime, value, vt.statsExpFactor)
+}
+
+type RequestStatsItem struct {
+	Name                string
+	ReqAmount, ReqValue float64
+}
+
+// RequestStats returns the current contents of the reference request basket, with
+// request values meaning average per request rather than total.
+func (vt *ValueTracker) RequestStats() []RequestStatsItem {
+	vt.statsExpLock.RLock()
+	expFactor := vt.statsExpFactor
+	vt.statsExpLock.RUnlock()
+	vt.lock.Lock()
+	defer vt.lock.Unlock()
+
+	vt.periodicUpdate()
+	res := make([]RequestStatsItem, len(vt.refBasket.basket.items))
+	for i, item := range vt.refBasket.basket.items {
+		res[i].Name = vt.mappings[vt.currentMapping][i]
+		res[i].ReqAmount = expFactor.Value(float64(item.amount)/basketFactor, vt.refBasket.basket.exp)
+		res[i].ReqValue = vt.refBasket.reqValues[i]
+	}
+	return res
+}
+
+// TotalServiceValue returns the total service value provided by the given node (as
+// a function of the weights which are calculated from the request timeout value).
+func (vt *ValueTracker) TotalServiceValue(nv *NodeValueTracker, weights ResponseTimeWeights) float64 {
```

Comment:
If it's an unused function, please delete it

Reply:
It is used by the server pool which is the subsequent PR.